### PR TITLE
Add option types

### PR DESCRIPTION
**Pull request by trapacska**

### CODE REVIEW

File: models/option.go [hunk 1]
```diff
@@ -5,10 +5,42 @@ import (
...
+// Type is to select the user interaction type that is required to fill an option
+type Type string
+
+// OptionTypes we currently support, list, user-input, optional-user-input
+const (
+	// Originally:
+	// - if there was only one key then this question was not asked
+	// - if there was more than one keys then it was asked in a list to select from
+	// Now, if this type is selected:
+	// - if there is only one key then this question shall not not be asked
```
Comment (BirmacherAkos) [inline models/option.go]: Note:
Comment typo:
`shall not not be asked`

File: models/option.go [hunk 2]
```diff
@@ -61,6 +93,12 @@ func (option *OptionNode) IsEmpty() bool {
 	return !option.IsValueOption() && !option.IsConfigOption()
 }
 
+// SetType ...
+func (option *OptionNode) SetType(optionType Type) *OptionNode {
```
Comment (BirmacherAkos) [inline models/option.go]: Note:
If the function already returns the OptionNode at the end, then you could use a value receiver instead.

Comment (trapacska) [reply models/option.go]: the reason I left it like this because then use it in a chained way and a single set way, 
```
myoption.SetType(...).DoOtherThing()
but also
myoption.SetType(...)
myoption.DoOtherThing()
```

File: scanner/config.go [hunk 3]
```diff
@@ -168,6 +168,33 @@ func runScanners(scannerList []scanners.ScannerInterface, searchDir string) map[
 	return scannerOutputs
 }
 
+func initOptionTypeRecursive(option *models.OptionNode) {
```
Comment (lpusok) [inline scanner/config.go]: An alternative solution suggested is to set the default type selector when creating an option with NewOption. Manually Remove all occurences of "_" used as a key in optionNode map. (Besides the usages for fastlane project type (toolscanner), and check that also works after the changes)

File: models/option.go [hunk 4]
```diff
@@ -5,10 +5,42 @@ import (
...
+// OptionTypes we currently support, list, user-input, optional-user-input
+const (
+	// Originally:
+	// - if there was only one key then this question was not asked
+	// - if there was more than one keys then it was asked in a list to select from
+	// Now, if this type is selected:
+	// - if there is only one key then this question shall not not be asked
+	// - if there are more than one keys then the selection must be asked in a list to select from
+	TypeSelector Type = "selector"
+	// Originally:
```
Comment (lpusok) [inline models/option.go]: If possible, update the comments to focus on the current expected behaviour (as historical one can be found in the git history)

File: vendor/github.com/bitrise-io/goinp/goinp/goinp.go [hunk 5]
```diff
@@ -58,6 +62,95 @@ func AskForString(messageToPrint string) (string, error) {
...
+		log.Printf("[%d] : %s", i+1, option)
+	}
+
+	for {
+		fmt.Print("Type in the option's number, then hit Enter: ")
+
+		answer, err := bufio.NewReader(os.Stdin).ReadString('\n')
+		if err != nil {
+			log.Errorf("failed to read input value")
+			continue
```
Comment (lpusok) [inline vendor/github.com/bitrise-io/goinp/goinp/goinp.go]: Is it a better way to return with err on failure in this case?

File: vendor/github.com/bitrise-io/goinp/goinp/goinp.go [hunk 6]
```diff
@@ -58,6 +62,95 @@ func AskForString(messageToPrint string) (string, error) {
...
+			log.Errorf("failed to parse option number, pick a number from 1-%d", len(options))
+			continue
+		}
+
+		if optionNo-1 < 0 || optionNo-1 >= len(options) {
+			log.Errorf("invalid option number, pick a number 1-%d", len(options))
+			continue
+		}
+
+		if options[optionNo-1] == customValueOptionText {
```
Comment (lpusok) [inline vendor/github.com/bitrise-io/goinp/goinp/goinp.go]: It would ber preferrable to use `(optionsNo ==len(options) && optional)` instead of comparing to the text value itself.

File: vendor/github.com/bitrise-io/goinp/goinp/goinp.go [hunk 7]
```diff
@@ -8,6 +8,10 @@ import (
 	"os"
 	"strconv"
 	"strings"
+	"syscall"
```
[the remark this answers is not in the document]
Comment (trapacska) [reply vendor/github.com/bitrise-io/goinp/goinp/goinp.go]: done 👍

File: vendor/github.com/bitrise-io/goinp/goinp/goinp.go (hunk 5 above)
Comment (trapacska) [reply vendor/github.com/bitrise-io/goinp/goinp/goinp.go]: I think this is better this way, especially if the user already answered N questions to not to fail because of a wrong input

File: scanner/utils.go [hunk 8]
```diff
@@ -13,35 +14,29 @@ import (
 )
 
 func askForOptionValue(option models.OptionNode) (string, string, error) {
```
Comment (godrei) [inline scanner/utils.go]: pls create a separate task to support the new input types in our interactive tools, including the scanner itself

File: models/option.go [hunk 9]
```diff
@@ -61,6 +94,12 @@ func (option *OptionNode) IsEmpty() bool {
 	return !option.IsValueOption() && !option.IsConfigOption()
 }
 
+// SetType ...
+func (option *OptionNode) SetType(optionType Type) *OptionNode {
```
Comment (godrei) [inline models/option.go]: Pls set the type in the Option initializer (`NewOption`) to avoid issues where we do not set type accidentally

File: scanner/utils.go [hunk 10]
```diff
@@ -13,30 +14,88 @@ import (
...
+	return
+}
+
+func selectOption(options []string) (string, error) {
+	for i, option := range options {
+		fmt.Printf("[%d] : %s\n", i+1, option)
+	}
+	fmt.Printf("Type in the option's number, then hit Enter: ")
+
+	answer, err := goinp.AskForOptionalInput("", false)
```
Comment (godrei) [inline scanner/utils.go]: NOTE: pretty bed client-side interface: `AskForOptionalInput` optional = false.

File: scanner/utils.go [hunk 11]
```diff
@@ -71,11 +130,17 @@ func AskForOptions(options models.OptionNode) (string, []envmanModels.Environmen
...
 		} else {
 			// go to the next option, based on the selected value
-			childOptions, found := opt.ChildOptionMap[selectedValue]
+			childOption, found := opt.ChildOptionMap[selectedValue]
 			if !found {
-				return nil
+				if opt.Type != models.TypeOptionalSelector {
+					return nil
+				}
+				for _, option := range opt.ChildOptionMap {
```
Comment (godrei) [inline scanner/utils.go]: pls leave a comment where you explain why is this case here

File: scanner/utils.go [hunk 12]
```diff
@@ -13,30 +14,88 @@ import (
 	"github.com/bitrise-io/goinp/goinp"
 )
 
+func getDefaultValue(opt models.OptionNode) string {
+	if opt.Type == models.TypeOptionalSelector {
```
Comment (godrei) [inline scanner/utils.go]: why do not we provide a default value in case of optional selector?

File: scanners/android/android.go [hunk 13]
```diff
@@ -91,16 +91,109 @@ func (scanner *Scanner) Options() (models.OptionNode, models.Warnings, models.Ic
 
 // DefaultOptions ...
 func (scanner *Scanner) DefaultOptions() models.OptionNode {
-	projectLocationOption := models.NewOption(ProjectLocationInputTitle, ProjectLocationInputEnvKey)
-	moduleOption := models.NewOption(ModuleInputTitle, ModuleInputEnvKey)
-	variantOption := models.NewOption(VariantInputTitle, VariantInputEnvKey)
-	configOption := models.NewConfigOption(DefaultConfigName, nil)
+	// projectLocationOption := models.NewOption(ProjectLocationInputTitle, ProjectLocationInputEnvKey).SetType(models.TypeUserInput)
```
Comment (godrei) [inline scanners/android/android.go]: please revert these changes and add the option type input to the option initialiser function

File: scanner/utils.go [hunk 14]
```diff
@@ -12,36 +14,88 @@ import (
...
+}
+
+func getOptions(opts map[string]*models.OptionNode) (options []string) {
+	for key := range opts {
+		options = append(options, key)
+	}
+	return
+}
+
+func selectOption(options []string) (string, error) {
```
Comment (godrei) [inline scanner/utils.go]: pls restore the previous auto select feature in case of 1 element lists

File: scanner/utils.go (hunk 10 above)
Comment (trapacska) [reply scanner/utils.go]: NOTE: this comment supposed to be in the package PR review

File: scanners/reactnative/expo.go [hunk 15]
```diff
@@ -183,35 +183,35 @@ entries.`
...
 		for _, exportMethod := range ios.IosExportMethods {
 			exportMethodOption.AddOption(exportMethod, workDirOption)
 		}
 
-		projectLocationOption := models.NewOption(android.ProjectLocationInputTitle, android.ProjectLocationInputEnvKey)
+		projectLocationOption := models.NewOption(android.ProjectLocationInputTitle, android.ProjectLocationInputEnvKey, models.TypeSelector)
 		workDirOption.AddOption(relPackageJSONDir, projectLocationOption)
 
-		moduleOption = models.NewOption(android.ModuleInputTitle, android.ModuleInputEnvKey).SetType(models.TypeUserInput)
+		moduleOption = models.NewOption(android.ModuleInputTitle, android.ModuleInputEnvKey, models.TypeUserInput)
```
Comment (godrei) [inline scanners/reactnative/expo.go]: should not it be a selector with a single value `app`?

File: scanners/reactnative/expo.go [hunk 16]
```diff
@@ -183,35 +183,35 @@ entries.`
...
 		projectLocationOption.AddOption(filepath.Join(relPackageJSONDir, "android"), moduleOption)
 	}
 
-	buildVariantOption := models.NewOption(android.VariantInputTitle, android.VariantInputEnvKey).SetType(models.TypeOptionalUserInput)
+	buildVariantOption := models.NewOption(android.VariantInputTitle, android.VariantInputEnvKey, models.TypeOptionalUserInput)
```
Comment (godrei) [inline scanners/reactnative/expo.go]: should not it be a selector with a single value `Release`?

File: scanners/reactnative/expo.go [hunk 17]
```diff
@@ -442,40 +442,40 @@ func (scanner *Scanner) expoConfigs() (models.BitriseConfigMap, error) {
...
-		projectLocationOption := models.NewOption(android.ProjectLocationInputTitle, android.ProjectLocationInputEnvKey)
+		projectLocationOption := models.NewOption(android.ProjectLocationInputTitle, android.ProjectLocationInputEnvKey, models.TypeSelector)
 		workDirOption.AddOption("", projectLocationOption)
 
-		moduleOption := models.NewOption(android.ModuleInputTitle, android.ModuleInputEnvKey).SetType(models.TypeUserInput)
+		moduleOption := models.NewOption(android.ModuleInputTitle, android.ModuleInputEnvKey, models.TypeUserInput)
 		projectLocationOption.AddOption("./android", moduleOption)
 
-		buildVariantOption := models.NewOption(android.VariantInputTitle, android.VariantInputEnvKey).SetType(models.TypeOptionalUserInput)
+		buildVariantOption := models.NewOption(android.VariantInputTitle, android.VariantInputEnvKey, models.TypeOptionalUserInput)
```
Comment (godrei) [inline scanners/reactnative/expo.go]: same here

File: scanners/reactnative/expo.go [hunk 18]
```diff
@@ -442,40 +442,40 @@ func (scanner *Scanner) expoConfigs() (models.BitriseConfigMap, error) {
...
 		for _, exportMethod := range ios.IosExportMethods {
 			exportMethodOption.AddOption(exportMethod, workDirOption)
 		}
 
-		projectLocationOption := models.NewOption(android.ProjectLocationInputTitle, android.ProjectLocationInputEnvKey)
+		projectLocationOption := models.NewOption(android.ProjectLocationInputTitle, android.ProjectLocationInputEnvKey, models.TypeSelector)
 		workDirOption.AddOption("", projectLocationOption)
 
-		moduleOption := models.NewOption(android.ModuleInputTitle, android.ModuleInputEnvKey).SetType(models.TypeUserInput)
+		moduleOption := models.NewOption(android.ModuleInputTitle, android.ModuleInputEnvKey, models.TypeUserInput)
```
Comment (godrei) [inline scanners/reactnative/expo.go]: same here
